tests/test_config/test_config_rules.py

```diff
@@ -0,0 +1,728 @@
+        "Three_Hours, Twelve_Hours, TwentyFour_Hours}" in err["Message"]
+    )
+
+    managed_rule = managed_config_rule()
+    managed_rule["ConfigRuleState"] = "BOGUS"
+    with pytest.raises(ClientError) as exc:
+        client.put_config_rule(ConfigRule=managed_rule)
+    err = exc.value.response["Error"]
+    assert err["Code"] == "ValidationException"
+    assert (
+        "Value 'BOGUS' at 'configRule.configRuleState' failed to satisfy "
+        "constraint: Member must satisfy enum value set: {ACTIVE, "
+        "DELETING, DELETING_RESULTS, EVALUATING}" in err["Message"]
+    )
+
+    managed_rule = managed_config_rule()
+    managed_rule["ConfigRuleState"] = "DELETING"
+    with pytest.raises(ClientError) as exc:
+        client.put_config_rule(ConfigRule=managed_rule)
+    err = exc.value.response["Error"]
+    assert err["Code"] == "InvalidParameterValueException"
+    assert (
+        "The ConfigRuleState DELETING is invalid.  Only the following values "
+        "are permitted: ACTIVE" in err["Message"]
+    )
+
+    managed_rule = managed_config_rule()
+    managed_rule["CreatedBy"] = "tester"
+    with pytest.raises(ClientError) as exc:
+        client.put_config_rule(ConfigRule=managed_rule)
+    err = exc.value.response["Error"]
+    assert err["Code"] == "InvalidParameterValueException"
+    assert (
+        "AWS Config populates the CreatedBy field for ServiceLinkedConfigRule. "
+        "Try again without populating the CreatedBy field" in err["Message"]
+    )
+
+
+@mock_config
+def test_aws_managed_rule_errors():
+    """Test various error conditions in ConfigRule instantiation."""
+    client = boto3.client("config", region_name=TEST_REGION)
+
+    # Extra, unknown input parameter should raise an error.
+    managed_rule = managed_config_rule()
+    managed_rule["Source"]["SourceIdentifier"] = "IAM_PASSWORD_POLICY"
+    managed_rule["InputParameters"] = '{"RequireNumbers":"true","Extra":"10"}'
+    with pytest.raises(ClientError) as exc:
+        client.put_config_rule(ConfigRule=managed_rule)
+    err = exc.value.response["Error"]
+    assert err["Code"] == "InvalidParameterValueException"
+    assert (
+        f"Unknown parameters provided in the inputParameters: "
+        f"{managed_rule['InputParameters']}" in err["Message"]
+    )
+
+    # Missing required parameters should raise an error.
+    managed_rule = managed_config_rule()
+    managed_rule["Source"]["SourceIdentifier"] = "CLOUDWATCH_ALARM_ACTION_CHECK"
+    with pytest.raises(ClientError) as exc:
+        client.put_config_rule(ConfigRule=managed_rule)
+    err = exc.value.response["Error"]
+    assert err["Code"] == "InvalidParameterValueException"
+    assert (
+        "The required parameter [alarmActionRequired, "
+        "insufficientDataActionRequired, okActionRequired] is not present "
+        "in the inputParameters" in err["Message"]
+    )
+
+    # If no MaxExecutionFrequency specified, set it to the default.
```

Comment:
Looks like this is now covered in `test_valid_put_config_custom_rule`?

Reply:
There are two `MaximumExecutionFrequency`, one at the `ConfigRule.Source.SourceDetails` level and one at the `ConfigRule` level.  I am very confused as to how they're related and how to set the default values.  I raised a question in the AWS Forum and never got an answer.  I experimented a bit and between boto3 and the AWS Console, it was no clearer how they interacted.

What's even more confusing is that for a custom rule, you can have multiple rules and they can each have a different  `MaximumExecutionFrequency` value, so how do you determine the value for the one at the ConfigRule level?

This will probably have to be revisited when people complain it's not working as expected or if I ever get an answer from AWS.

In any case, `test_aws_managed_rule_errors()` is testing at the `ConfigRule.Source.SourceDetails` level for AWS managed rules, and `test_valid_put_config_custom_rule` is testing `ConfigRule.Source.SourceDetails` level for custom rules. 

All very messy.

Reply:
Ah, gotcha. Thanks for the explanation.